test/unit/assorted/server_discovery_and_monitoring.spec.test.ts

```diff
@@ -311,6 +314,102 @@ const SDAM_EVENTS = [
   ...HEARTBEAT_EVENTS
 ];
 
+function makeConnectionsError(appError: ApplicationError): sinon.SinonStub[] {
```

Comment:
Sorry for the large diff here, reviewers.  Our test runner's old mocks were too permissive and allowed tests to pass that should have failed.  I've adjusted this mocking logic:

- if we need to stub before handshake errors, I create stubs that provide the bare minimum needed for the actual handshake to fail.  In practice, this means stubbing `net.createConnection()`.
- otherwise, I use rely on the old stub logic.

I've also adjusted the code to use `switch` statements instead of `if-statements` when switching on discriminated union fields, because TS errors if we haven't exhausted all switch statements.